src/services/MessageService.ts

```diff
@@ -0,0 +1,68 @@
+import { singleton } from 'tsyringe';
+import Message, { APIMessage } from '../entities/Message';
+import { getRepository } from 'typeorm';
+import { APIError, formatValidationErrors } from '../util/errors';
+import { validateOrReject } from 'class-validator';
+import { Channel } from '../entities/Channel';
+
+type MessageCreationData = Omit<APIMessage, 'id'>;
+
+enum GetMessageError {
+	NotFound = 'Message not found',
+	InvalidChannel = 'Message does not exist for given channel'
+}
+
+enum DeleteMessageError {
+	NotFound = 'Message not found',
+	InvalidChannel = 'Message does not exist for given channel',
+	NotAuthor = 'You are not the author of this message'
+}
+
+enum GetMessagesError {
+	ChannelNotFound = 'Channel does not exist'
+}
+
+@singleton()
+export default class MessageService {
+	public async createMessage(data: MessageCreationData): Promise<APIMessage> {
+		const message = new Message();
+		message.content = data.content;
+		message.time = new Date(data.time);
+		message.author = { id: data.authorID } as any;
+		message.channel = { id: data.channelID } as any;
+		await validateOrReject(message).catch(e => Promise.reject(formatValidationErrors(e)));
+		return (await getRepository(Message).save(message)).toJSON();
+	}
+
+	public async getMessage(data: Pick<APIMessage, 'id' | 'channelID'>): Promise<APIMessage> {
+		if (!data.id) throw new APIError(404, GetMessageError.NotFound);
+		const message = await getRepository(Message).findOneOrFail(data.id).catch(() => Promise.reject(new APIError(404, GetMessageError.NotFound)));
+		if (message.channel.id !== data.channelID) throw new APIError(400, GetMessageError.InvalidChannel);
+		return message.toJSON();
+	}
+
+	public async getMessages(data: { channelID: string; page?: number; count: number }): Promise<APIMessage[]> {
+		if (!data.page || isNaN(data.page)) data.page = 0;
+		if (!data.channelID) throw new APIError(404, GetMessagesError.ChannelNotFound);
+		const channel = await getRepository(Channel).findOneOrFail(data.channelID).catch(() => Promise.reject(new APIError(404, GetMessagesError.ChannelNotFound)));
+		const messages = await getRepository(Message)
+			.createQueryBuilder('message')
+			.leftJoinAndSelect('message.channel', 'channel')
+			.leftJoinAndSelect('message.author', 'author')
```

Comment:
In the `Message` entity, both the `channel` and `author` entities are eagerly loaded. If you use the `find` method you can avoid doing the joining yourself. Both ways work, its up to you :)

Using`.find` method, it would be something along the lines of:
```typescript
  ... = await getRepository(Message)
  .find({
    where: {  channel: channel.id },
    order: { time: 'DESC' },
    skip: { data.count * data.page },
    take: { data.count }
  });
  
```

Reply:
Ahh nice!